File: price_compare_web/services/matcher.py
"""
Product matching using RapidFuzz.
Scoring: 60% text similarity + 25% size similarity + 15% keyword overlap.
"""
import re
from rapidfuzz import fuzz
from config import WEIGHT_TEXT, WEIGHT_SIZE, WEIGHT_KEYWORDS


def _extract_keywords(text: str) -> set[str]:
    """Lowercase alpha-numeric tokens, stop-words removed."""
    stop = {"the", "a", "an", "and", "or", "of", "with", "for", "in", "oz", "lb",
            "g", "kg", "ml", "l", "ct", "pk", "fl", "pack", "count"}
    return {w for w in re.findall(r"[a-z0-9]+", text.lower()) if w not in stop}


def _size_similarity(size_a: float | None, size_b: float | None) -> float:
    """Return 0-1 similarity between two normalized sizes."""
    if size_a is None or size_b is None:
        return 0.5  # neutral when unknown
    if size_a == 0 and size_b == 0:
        return 1.0
    larger  = max(size_a, size_b)
    smaller = min(size_a, size_b)
    if larger == 0:
        return 1.0
    return smaller / larger


def score_match(
    query: str,
    candidate_title: str,
    query_size: float | None,
    candidate_size: float | None,
) -> float:
    """Return composite match score in [0, 100]."""
    text_score = fuzz.token_set_ratio(query, candidate_title) / 100.0

    size_score = _size_similarity(query_size, candidate_size)

    q_kw  = _extract_keywords(query)
    c_kw  = _extract_keywords(candidate_title)
    if q_kw | c_kw:
        kw_score = len(q_kw & c_kw) / len(q_kw | c_kw)
    else:
        kw_score = 0.0

    composite = (
        WEIGHT_TEXT     * text_score +
        WEIGHT_SIZE     * size_score +
        WEIGHT_KEYWORDS * kw_score
    ) * 100.0

    return round(composite, 2)


def find_best_match(
    query: str,
    query_size: float | None,
    candidates: list[dict],  # each has 'title', 'quantity' (normalized)
) -> tuple[dict | None, float]:
    """
    Score all candidates and return (best_candidate, score).
    Returns (None, 0.0) if candidates is empty.
    """
    if not candidates:
        return (None, 0.0)

    scored = []
    for c in candidates:
        s = score_match(
            query,
            c.get("title", ""),
            query_size,
            c.get("quantity"),
        )
        scored.append((s, c))

    scored.sort(key=lambda x: x[0], reverse=True)
    best_score, best = scored[0]
    return (best, best_score)

File: docs/js/matcher.js
/**
 * Product matching — fuzzy text + size + keyword scoring.
 * Uses a simple Levenshtein-based token set similarity (no external library).
 */

// Simple Levenshtein distance
function levenshtein(a, b) {
  const m = a.length, n = b.length;
  const dp = Array.from({ length: m + 1 }, () => new Array(n + 1).fill(0));
  for (let i = 0; i <= m; i++) dp[i][0] = i;
  for (let j = 0; j <= n; j++) dp[0][j] = j;
  for (let i = 1; i <= m; i++)
    for (let j = 1; j <= n; j++)
      dp[i][j] = a[i - 1] === b[j - 1]
        ? dp[i - 1][j - 1]
        : 1 + Math.min(dp[i - 1][j], dp[i][j - 1], dp[i - 1][j - 1]);
  return dp[m][n];
}

// Token set ratio: compare sorted unique token intersection+remainder
function tokenSetRatio(a, b) {
  const ta = [...new Set(a.toLowerCase().split(/\s+/).filter(Boolean))].sort();
  const tb = [...new Set(b.toLowerCase().split(/\s+/).filter(Boolean))].sort();
  const sa = ta.join(" ");
  const sb = tb.join(" ");
  const maxLen = Math.max(sa.length, sb.length);
  if (maxLen === 0) return 100;
  return ((1 - levenshtein(sa, sb) / maxLen) * 100);
}

const STOP_WORDS = new Set([
  "the", "a", "an", "and", "or", "of", "with", "for", "in",
  "oz", "lb", "g", "kg", "ml", "l", "ct", "pk", "fl", "pack", "count"
]);

function extractKeywords(text) {
  return new Set(
    text.toLowerCase().match(/[a-z0-9]+/g)?.filter(w => !STOP_WORDS.has(w)) || []
  );
}

function sizeSimilarity(a, b) {
  if (a == null || b == null) return 0.5;
  if (a === 0 && b === 0) return 1;
  const larger = Math.max(a, b);
  const smaller = Math.min(a, b);
  return larger === 0 ? 1 : smaller / larger;
}

const W_TEXT = 0.60, W_SIZE = 0.25, W_KW = 0.15;

export function scoreMatch(query, candidateTitle, querySize, candidateSize) {
  const textScore = tokenSetRatio(query, candidateTitle) / 100;
  const sizeScore = sizeSimilarity(querySize, candidateSize);

  const qKw = extractKeywords(query);
  const cKw = extractKeywords(candidateTitle);
  const union = new Set([...qKw, ...cKw]);
  const inter = new Set([...qKw].filter(x => cKw.has(x)));
  const kwScore = union.size ? inter.size / union.size : 0;

  return +((W_TEXT * textScore + W_SIZE * sizeScore + W_KW * kwScore) * 100).toFixed(2);
}
